File: lib/cordova/file_merger.rb
require 'fileutils'
require 'rubygems'
require 'json'

class FileMerger
  def initialize
  end

  def run(filepath)
    # skip the guides
    return if filepath.match(/\/guide\//)
    
    # skip missing files (file that are merged are also deleted)
    return unless File.exists?(filepath)
    
    # file info
    @filename  = File.basename(filepath)
    @directory = File.dirname(filepath)
    
    # skip unless file is referenced in the merge JSON
    return unless config.include?(@filename)
    
    # open the file to merge into
    File.open filepath, 'a' do |f|
      # loop over the files to merge
      config[@filename].each do |filepath|
        # skip the file that we're merging into because it's listed in config.json
        next if File.basename(filepath) == @filename
        
        # hacky to qualify the path
        filepath = File.join('tmp', 'docs', filepath)
        
        # append and delete the file
        f.write "\n\n---\n"
        f.write File.read(filepath).strip
        FileUtils.rm filepath
      end
    end
  end

  def config
      return @config unless @config.nil?
      
      directory = @directory
      
      while @config.nil?
        file = File.join(directory, 'config.json')
        
        if File.exists?(file)
          @config = (JSON.parse IO.read(file))['merge']
        else
          directory = File.dirname(directory)
        end
      end
      
      return @config
  end

  def file_exists?(file_path)
    return true if File.exists? file_path

    puts "Missing: #{file_path}"
    return false
  end
end

File: tools/lib/cordova/pre/filemerger.js
/*
       Licensed to the Apache Software Foundation (ASF) under one
       or more contributor license agreements.  See the NOTICE file
       distributed with this work for additional information
       regarding copyright ownership.  The ASF licenses this file
       to you under the Apache License, Version 2.0 (the
       "License"); you may not use this file except in compliance
       with the License.  You may obtain a copy of the License at

         http://www.apache.org/licenses/LICENSE-2.0

       Unless required by applicable law or agreed to in writing,
       software distributed under the License is distributed on an
       "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
       KIND, either express or implied.  See the License for the
       specific language governing permissions and limitations
       under the License.
*/
/*jslint node: true */
var fs = require("fs-extra"),
    path = require("path"),
    FileHelpers = require("../../file_helpers");

var FileMerger = (function () {
    'use strict';

    var config = null;

    /**
    * Creates a new instance of FileMerger
    * @param options Options for the generation process.
    */
    function FileMerger(options) {
        this.options = options || { verbose: 0 };
        this.stage = "Merge files";

        config = null;
    }

    FileMerger.prototype.run = function (file) {
        if (this.options.verbose > 1) {
            console.log("Merge file " + file);
        }

        if (file.match(/\/guide\//)) {
            return null;
        }

        if (!fs.existsSync(file)) {
            return null;
        }

        this.filename  = path.basename(file);
        this.directory = path.dirname(file);

        var filesToAppend,
            cfg = this.config(),
            i,
            filepath;
        if (!cfg.hasOwnProperty(this.filename)) {
            return null;
        }

        filesToAppend = cfg[this.filename];
        for (i = 0; i < filesToAppend.length; i += 1) {
            filepath = filesToAppend[i];

            // skip the file that we're merging into because it's listed in config.json
            if (path.basename(filepath) === this.filename) {
                if (this.options.verbose > 0) {
                    console.log("Skipping " + filesToAppend[i] + " from " + this.filename + " section since it has same base name as file to merge");
                }
            } else {
                filepath = path.join(FileHelpers.getTmpDirectory(), 'docs', filepath);
                fs.appendFileSync(file, "\n\n---\n");
                if (!fs.existsSync(filepath)) {
                    console.log("File " + filepath + " is missing and thus don't merged with " + file);
                } else {
                    fs.appendFileSync(file, fs.readFileSync(filepath, "utf-8").trim());
                    fs.removeSync(filepath);
                }
            }
        }
    };

    FileMerger.prototype.config = function () {
        if (config !== null) {
            return config;
        }

        var directory = this.directory,
            file,
            configJSON;
        while (config === null) {
            file = path.join(directory, 'config.json');
            if (fs.existsSync(file)) {
                configJSON = fs.readFileSync(file);
                config = JSON.parse(configJSON).merge;
            } else {
                directory = fs.dirname(directory);
            }
        }

        return config;
    };

    return FileMerger;
}());
module.exports = FileMerger;
